src/ir/effects.h

```diff
@@ -527,6 +532,11 @@ struct EffectAnalyzer
     }
     return effects;
   }
+
+  void ignoreBranches() {
+    branchesOut = false;
+    breakTargets.clear();
```

Comment:
That's a very good point. Yeah, this looks odd atm, and I think we should change it.

I think it would be risky to reset `throws` since it's a particular type of side effect. It counts as a "global" side effect in particular since its observable from outside the module. I worry resetting that here would lead to bugs some day.

Alternatively, I think we can make `branches` only do `branchesOut || hasExternalBreakTargets()` - which is simpler and more intuitive. That does change the current behavior of some passes like CodePushing, which with that change can do this:
```diff
  (func $cant-push-past-call
   (local $x i32)
   (block $out
-   (local.set $x
-    (i32.const 1)
-   )
    (call $cant-push-past-call)
    (drop
     (i32.const 1)
    )
    (br_if $out
     (i32.const 2)
    )
+   (local.set $x
+    (i32.const 1)
+   )
    (drop
     (local.get $x)
    )
   )
  )
```
That is, it can move the set past the call. The call might throw, but it ignores that. I believe this is safe: if we throw all the way out of the function, we don't need to care about local state.

Correct me if I'm wrong, but I think that that is what `throws` means - it's set when we are *not* in a catch scope (a catch scope would allow control flow to continue in the function). So if it is set, we can't branch inside the function, and CodePushing etc. are ok.

If that's right, I can do that as a followup PR? (seems safer separate since before this PR we had `bool transfersControlFlow() const { return branches || throws; }`, that is, `throws` being taken into account there is not changed in this PR).

Reply:
Oh, no, I don't think we can exclude `throws` here. `throws` means simply the expression contains a call that can throw. We don't know whether it is within a scope that will be caught, and it's not possible to figure out in an expression level anyway. So if some expression can throw, it can be caught within the function and execution can continue in the same function. So I think `transfersControlFlow` should be as is, and I'm also little unsure it's better to rename it to `branches()`, because branches intuitively means just branches and not throws (or returns).

Reply:
Oh, you're right, yeah... I was confused.

It would be nice to optimize that kind of thing eventually, but it does need to look at all the outer scopes in the function.

And I agree the renaming here to `branches` is not that good, given that. I'll return it to `transfersControlFlow`.

Reply:
One thing to note here is `throws` is not set if it is already resolved within an expression. For example:
```wast
(try
  (do (call $foo)) ; can throw
  (catch ...)
)
```
This expression does not set `throws` to true because it is already shown than this call is caught within this expression. But it sets `throws` to true if it contains a call that can exist this expression (whether it will be caught within this function or not, which we can't figure out with this expression only).